trieste/models/optimizer.py

```diff
@@ -0,0 +1,118 @@
+from typing import Any, Dict, Optional, Callable, Union, Iterable, Tuple
+from functools import singledispatch
+from dataclasses import dataclass, field
+import tensorflow as tf
+import gpflow
+
+from gpflow.models import ExternalDataTrainingLossMixin, InternalDataTrainingLossMixin
+from ..utils import jit
+from ..datasets import Dataset
+
+
+DatasetBuilder = Callable[[Dataset, Optional[int]], Union[Iterable, Tuple]]
+"""
+Type alias for a function that creates minibatches from a :class:`~optiflow.datasets.Dataset`.
+"""
+
+
+@dataclass
+class Optimizer:
+    optimizer: Union[gpflow.optimizers.Scipy, tf.optimizers.Optimizer]
+    minimize_args: Dict[str, Any] = field(default_factory=lambda: {})
+    compile: bool = False
+
+    def create_loss(self, model: tf.Module, dataset: Dataset) -> Callable:
```

Comment:
can this be private?

Reply:
Leave it public.

Reply:
@joelberkeley ^^^

Reply:
we will need tests for all public functions